### PR TITLE
uv for CI/CD

### PR DESCRIPTION
**Pull request by JamesABaker**
# Pull Request

## Description

Following discussion on #14, there is a community preference for uv over pip (further detailed in the adr in this PR).

Closes 
- #14 

## Checklist for submission

- [x] GitHub Actions are passing
- [x] I have added relevant tests (or explain why tests aren't needed)
- [x] I have considered if an ADR is required for any architectural decisions
- [x] Documentation has been updated if needed (README, comments, etc.)
- [x] The PR does not exceed 500 lines of code (excluding config files or data changes)

## To the reviewer

When doing a review, please follow [the GDS approach](https://gds-way.digital.cabinet-office.gov.uk/manuals/code-review-guidelines.html#how-to-review-code).

For this PR, please focus on:

- [ ] **Code Audit**. A comprehensive review according to the [ONS DuckBook standards](https://best-practice-and-impact.github.io/qa-of-code-guidance/checklist_lower.html#template-checklist)
- [ ] **Peer Review**. A review of the process, method, or maths being sensible
- [x] **Code Review**. A code review for:
    - code implementation (does the code do what it says it does)
    - documentation (is it easy to understand what the code is trying to do)
    - tests (is the code being tested for edge cases, unexpected inputs, and normal operation)

### CODE REVIEW

File: .github/workflows/pytest.yml [hunk 1]
```diff
@@ -11,15 +11,16 @@ jobs:
...
         with:
           python-version: "3.12"
-          cache: "pip"
+
+      - name: Install uv
+        uses: astral-sh/setup-uv@v5
+        with:
+          enable-cache: true
 
       - name: Install dependencies
```
Comment (Nathan-Williams-1) [inline .github/workflows/pytest.yml]: Typically when using `uv` we rely on a `uv.lock` file for dependency management. Should we check for that first to prefer the locked environment? The ordering could be:

- if `uv.lock` -> `uv sync --dev`
- otherwise, fall back to checking for requirements files and using uv pip install

That would give us more reproducible installs in repos that use `uv` natively.

File: .github/workflows/pytest.yml [hunk 2]
```diff
@@ -11,15 +11,16 @@ jobs:
...
         run: |
-          pip install --upgrade pip
-          pip install pytest pytest-cov
-          if [ -f requirements.txt ]; then pip install -r requirements.txt; fi
-          if [ -f requirements-dev.txt ]; then pip install -r requirements-dev.txt; fi
+          uv pip install --system -r requirements.txt
+          uv pip install --system -r requirements-dev.txt
```
[the remark this answers is not in the document]
Comment (JamesABaker) [reply .github/workflows/pytest.yml]: I did think about how to handle this and thought the user should get an error if one of them doesn't exist. Perhaps that is only true for the main requirements.txt.

Out of the box, there is a requirements-dev.txt that includes pytest, so the developers using this repo can decide if they like that way of working or not.

[e344853](https://github.com/moj-analytical-services/data-science-template/pull/31/commits/e344853a88c737105953cf83dd994307b103d87a) first tries the uv.lock, then the requirements (will fail if not found) and then the requirements-dev.
